### DISCUSSION TITLE
Default Values not working as expected

### DISCUSSION THREAD

**Discussion by vikas990** [react-hook-form/react-hook-form#7058] [Q&A] [answered]
hi, I m facing an issue related to default values as I set the data which is coming from the API and set it to the default values . I m also resetting the values but I can't see the values on the fields and the same problem for dropdowns also. i m using Next js and Material UI for the frontend.

# Use Form
<img width="760" alt="Screenshot 2021-11-16 at 1 37 43 PM" src="https://user-images.githubusercontent.com/50900838/141957621-df328b34-e40d-4179-b8a2-0c3c804c352b.png">

# Input Fields
<img width="512" alt="Screenshot 2021-11-16 at 2 42 33 PM" src="https://user-images.githubusercontent.com/50900838/141957628-e5cbe506-b7b6-4acf-b509-0a59fa9c942c.png">

**Comment by JamesWong1999** [comment]
provide a codesandbox?

**Reply by vikas990** [reply]
hy, thank you for the reply i m using this code to apply defaultValues 

const ProfileComponent = ({
  onImageChange,
  imageUrl,
  useForm,
  SubmitHandler,
  companyData,
  onChange,
}) => {
  const {
    register,
    handleSubmit,
    formState: { errors },
    reset,
  } = useForm({
    resolver: yupResolver(schema),
    defaultValues: {
      name: companyData?.name,
      about: companyData?.about,
      organizationType: companyData?.organizationType,
    },
  });
useEffect(() => {
    reset({
      name: companyData?.name,
      about: companyData?.about,
      organizationType: companyData?.organizationType,
    });
  }, [reset, companyData]);

  return (
    <>
      <Card>
        <LeftContainer>
          {imageUrl ? (
            <Heading>
              <Image
                src={imageUrl}
                alt="Image of company logo"
                height="100%"
                width="100%"
              />
              <p>Company Logo</p>
            </Heading>
          ) : (
            <Heading>
              <AccountCircleIcon fontSize="large" />
              <div>Profile</div>
            </Heading>
          )}
        </LeftContainer>
        <div>
          <RightContainerHeading>Enter Company Details</RightContainerHeading>
          <hr />
          <form onSubmit={handleSubmit(SubmitHandler)}>
            {companyData?.official_name ? (
              <InputSection>
                <InputLabel>
                  Company Brand Name <span>*</span>
                </InputLabel>
                <InputField
                  name="officialName"
                  type="text"
                  placeholder="E.g. Uber"
                  fullWidth
                  {...register("officialName")}
                  error={errors?.officialName?.message && true}
                  helperText={errors?.officialName?.message}
                />
              </InputSection>
            ) : (
              <InputSection>
                <InputLabel>
                  Company Brand Name <span>*</span>
                </InputLabel>
                <InputField
                  name="name"
                  type="text"
                  placeholder="E.g. Uber"
                  fullWidth
                  {...register("name")}
                  error={errors?.name?.message && true}
                  helperText={errors?.name?.message}
                />
              </InputSection>
            )}
            <InputSection>
              <InputLabel>
                Company Description <span>*</span>
              </InputLabel>
              <InputField
                name="about"
                type="text"
                placeholder="E.g. Works in health sector"
                rows={3}
                multiline
                fullWidth
                {...register("about")}
                error={errors?.about?.message && true}
                helperText={errors?.about?.message}
              />
            </InputSection>
            {!companyData.official_name && (
              <>
                <InputSection>
                  <InputLabel>
                    Organization Type <span>*</span>
                  </InputLabel>
                  <FormControl
                    error={errors?.organizationType?.message && true}
                    fullWidth
                  >
                    <Select
                      placeholder="Ex. StartUp"
                      name="organizationType"
                      {...register("organizationType")}
                    >
                      {degreeOptions.map((degreeOption) => (
                        <MenuItem key={uuidv4()} value={degreeOption?.value}>
                          {degreeOption?.text}
                        </MenuItem>
                      ))}
                    </Select>
                    <ErrorStyle>{errors?.organizationType?.message}</ErrorStyle>
                  </FormControl>
                </InputSection>
                <InputSection>
                  <InputLabel>
                    City <span>*</span>
                  </InputLabel>
                  <InputField
                    name="city"
                    type="text"
                    placeholder="E.g. Mumbai"
                    fullWidth
                    {...register("city")}
                    value={companyData?.city}
                    onChange={onChange}
                    error={errors?.city?.message && true}
                    helperText={errors?.city?.message}
                  />
                </InputSection>
                <InputSection>
                  <InputLabel>
                    Number Of Employees <span>*</span>
                  </InputLabel>

                  <InputField
                    name="numberOfEmployees"
                    type="number"
                    placeholder="E.g. 20"
                    fullWidth
                    {...register("numberOfEmployees")}
                    value={companyData?.numberOfEmployees}
                    onChange={onChange}
                    defaultValue=""
                    error={errors?.numberOfEmployees?.message && true}
                    helperText={errors?.numberOfEmployees?.message}
                  />
                </InputSection>
                <InputSection>
                  <InputLabel>
                    Address
                    <span>*</span>
                  </InputLabel>

                  <InputField
                    name="address"
                    type="text"
                    placeholder="E.g. 304,airoli plaza b"
                    fullWidth
                    {...register("address")}
                    value={companyData?.address}
                    onChange={onChange}
                    error={errors?.address?.message && true}
                    helperText={errors?.address?.message}
                  />
                </InputSection>
                <InputSection>
                  <InputLabel>
                    Industry
                    <span>*</span>
                  </InputLabel>
                  <FormControl
                    error={errors?.industry?.message && true}
                    fullWidth
                  >
                    <Select
                      placeholder="E.g. IT"
                      name="industry"
                      {...register("industry")}
                      defaultValue=""
                    >
                      {discOptions.map((discOption) => (
                        <MenuItem key={uuidv4()} value={discOption?.value}>
                          {discOption?.text}
                        </MenuItem>
                      ))}
                    </Select>
                  </FormControl>
                  <ErrorStyle> {errors?.industry?.message}</ErrorStyle>
                </InputSection>
                <InputSection>
                  <InputLabel>
                    Year of Registeration
                    <span>*</span>
                  </InputLabel>
                  <FormControl
                    error={errors?.yearOfIncorporation?.message && true}
                    fullWidth
                  >
                    <Select
                      placeholder="E.g 2020"
                      name="yearOfIncorporation"
                      fullWidth
                      {...register("yearOfIncorporation")}
                    >
                      {years.map((year) => (
                        <MenuItem key={uuidv4()} value={year?.value}>
                          {year?.text}
                        </MenuItem>
                      ))}
                    </Select>
                    <ErrorStyle>
                      {errors?.yearOfIncorporation?.message}
                    </ErrorStyle>
                  </FormControl>
                </InputSection>
              </>
            )}
            <InputSection>
              <InputLabel>
                Upload Company&apos;s logo (format: png/jpg/jpeg, max-size: 2
                MB)
              </InputLabel>

              <InputField
                name="file"
                type="file"
                hidden
                id="imageChange"
                fullWidth
                {...register("file")}
                onChange={onImageChange}
              />
            </InputSection>
            <InputSection>
              <InputLabel>Website</InputLabel>
              <InputField
                name="website"
                type="text"
                placeholder="E.g. www.google.com"
                fullWidth
                {...register("website")}
                value={companyData?.website}
                onChange={onChange}
              />
            </InputSection>
            <InputSection>
              <InputLabel>Linkdin</InputLabel>

              <InputField
                name="linkedIn"
                type="text"
                placeholder="E.g. www.LinkedIn.com"
                fullWidth
                {...register("linkedIn")}
                value={companyData?.url?.linkedIn}
                onChange={onChange}
              />
            </InputSection>

            <InputSection>
              <Button
                type="submit"
                bgcolor="secondary"
                color="secondary"
                fullWidth
              >
                Save and Continue
              </Button>
            </InputSection>
          </form>
        </div>
      </Card>
    </>
  );
};

thank you:-

**Comment by ritikbanger** [comment, marked as answer]
Hello, Default values will not work if you are using a custom component, though you can pass them using states like this:
```
const { user: user_temp } = useSelector((states: any) => states.authReducer)
const [user, setUser] = useState(user_temp)
```

i.e. here user_temp contains the default values.
and then use them like this:

```
const methods = useForm({
    defaultValues: useMemo(() => {
      return user
    }, [user]),
    resolver: yupResolver(validationSchema(translation)),
  })

  const {
    handleSubmit,
    reset,
    register,
    formState: { errors },
  } = methods
```

**Reply by vikas990** [reply]
I have also tried this but facing the same problem even after setting the reset. On the first render, the value is not showing in the fields but on the second render it is showing the values.

**Reply by ritikbanger** [reply]
Hello @vikas990 , the useStates take some time to set the states, so it will not be visible in the first render though you can set values initially like this:
`const [userInfo, setUserInfo] = useState(useSelector((states: any) => states.authReducer))` and in that field us this
`value={userInfo?.email}`

**Reply by vikas990** [reply]
hy @ritikbanger , if the defaultValues are not visible on the first render then it is a problem right and also if I set a value then I have to make a onChange function to change the values which should not be done if I m using React-hook-form.

**Reply by ritikbanger** [reply]
You can create an onChange function too even though you use React Hook Form. Call the function in every field or call it onSubmit.

**Reply by vikas990** [reply]
Still facing the same issue. on submit I can see the response of the preloaded data but it is not showing on the fields.

**Reply by ritikbanger** [reply]
I think you are using a custom input component InputField, try passing the register as a prop to your InputField like this: `register={register} `and in the file where you have defined the InputField give` {...register(name)}` there.

**Reply by vikas990** [reply]
But I think that's not a good approach if in the future I'll create a separate folder for my input fields it will cause a problem. Is there any other way to make default values work on custom components?

**Reply by ritikbanger** [reply]
No, passing props is the only method.

**Reply by vikas990** [reply]
I have recreated the bug.
CodeSandbox:-
https://codesandbox.io/s/summer-hooks-qtx7r

**Reply by vikas990** [reply]
Hy @ritikbanger ,  i have tried the solution and it works for the text field but not for the text field can please help me to understand what I m doing wrong.
https://codesandbox.io/s/summer-hooks-qtx7r?file=/components/Form.js

**Reply by ritikbanger** [reply]
Hello @vikas990, you are giving out of range values as your default value for dropdown.
1. You are using useEffect without dropdown value so it is reseting the value to ' ' and that's why it is not rendering your default value.
2. You are giving out of range select field value, since the field is a select field, it will not render out of range values, you can see a warning for the same.
I have tweaked a little and it has worked now. 

Here is the link: [https://codesandbox.io/s/agitated-snowflake-lx1ej?file=/components/Form.js](https://codesandbox.io/s/agitated-snowflake-lx1ej?file=/components/Form.js)


Here is the code without using useEffect and reset:

```
import { Card } from "./main.styles";
import { useForm, useWatch } from "react-hook-form";
import { useEffect } from "react";
import { Button, FormControl, InputField, MenuItem, Select } from "./utils";
const Controller = ({ control, register, name, rules, render }) => {
  const value = useWatch({
    control,
    name
  });
  const props = register(name);
  return render({
    value,
    onChange: (e) =>
      props.onChange({
        target: {
          name,
          value: e.target.value
        }
      }),

    onBlur: props.onBlur,
    name: props.name
  });
};
export default function Form() {
  const { register, handleSubmit, reset, control, setValue } = useForm({
    defaultValues: {
      name: "test",
      dropdown: "three"
    }
  });

  // useEffect(() => {
  //   reset({
  //     name: "test2",
  //     dropdown: "two"
  //   });
  // }, [reset]);
  const submitHandler = (data) => {
    console.log(data);
  };
  return (
    <div>
      <Card>
        <h1>Form</h1>
        <form onSubmit={handleSubmit(submitHandler)}>
          <InputField
            fullWidth
            placeholder="name"
            name="name"
            register={register}
          />
          <br />
          <br />
          <Controller
            {...{
              control,
              register,
              name: "dropdown",
              rules: {},
              render: (props) => (
                <FormControl fullWidth>
                  <Select {...props}>
                    <MenuItem value="one">one</MenuItem>
                    <MenuItem value="two">Two</MenuItem>
                    <MenuItem value="three">Three</MenuItem>
                  </Select>
                </FormControl>
              )
            }}
          />

          <br />
          <br />
          <Button fullWidth type="submit">
            Submit
          </Button>
        </form>
      </Card>
    </div>
  );
}

```

Here is code 2 with UseEffect and reset:
```

import { Card } from "./main.styles";
import { useForm, useWatch } from "react-hook-form";
import { useEffect } from "react";
import { Button, FormControl, InputField, MenuItem, Select } from "./utils";
const Controller = ({ control, register, name, rules, render }) => {
  const value = useWatch({
    control,
    name
  });
  const props = register(name);
  return render({
    value,
    onChange: (e) =>
      props.onChange({
        target: {
          name,
          value: e.target.value
        }
      }),

    onBlur: props.onBlur,
    name: props.name
  });
};
export default function Form() {
  const { register, handleSubmit, reset, control, setValue } = useForm({
    defaultValues: {
      name: "test",
      dropdown: "three"
    }
  });

  useEffect(() => {
    reset({
      name: "test2",
      dropdown: "two"
    });
  }, [reset]);
  const submitHandler = (data) => {
    console.log(data);
  };
  return (
    <div>
      <Card>
        <h1>Form</h1>
        <form onSubmit={handleSubmit(submitHandler)}>
          <InputField
            fullWidth
            placeholder="name"
            name="name"
            register={register}
          />
          <br />
          <br />
          <Controller
            {...{
              control,
              register,
              name: "dropdown",
              rules: {},
              render: (props) => (
                <FormControl fullWidth>
                  <Select {...props}>
                    <MenuItem value="one">one</MenuItem>
                    <MenuItem value="two">Two</MenuItem>
                    <MenuItem value="three">Three</MenuItem>
                  </Select>
                </FormControl>
              )
            }}
          />

          <br />
          <br />
          <Button fullWidth type="submit">
            Submit
          </Button>
        </form>
      </Card>
    </div>
  );
}
```

**Reply by vikas990** [reply]
hy @ritikbanger, thank you so much it works just great.

**Reply by vikas990** [reply]
yes, Thank you again.

**Comment by mbrowne** [comment]
I discovered that it's also possible to use a custom select component (that doesn't use a native `<select>`) without necessarily making it a controlled component. Here's an example (note: this is using a custom-built Select component, not material-ui, but I think a similar approach should work with material-ui or any other third-party Select component):

```
    const field = register('countryId')
    const initialVal = React.useRef(getValues(field.name))
    ...
    <Select
        {...field}
        defaultValue={initialVal.current}
        onChange={(selectedOption) => {
            setValue(field.name, selectedOption.value)
        }}
        options={countryOptions}
    />
```
